### DISCUSSION TITLE
How to handle Server Action network error?

### DISCUSSION THREAD

**Discussion by bravius** [vercel/next.js#58093] [Help]
### Summary

When a Server Action fails due to a network error, such as "ERR_INTERNET_DISCONNECTED", there seems to be no way to catch the error, specifically "TypeError: Failed to fetch," other than using global-error.tsx.

I've tried try/catch around the Server Action call, as well as error.js (App Router).

Using 14.0.2-canary.12

**Comment by tomjamesallen** [comment]
Any update on this? I've been testing out https://next-safe-action.dev/ and it's working really well for my use case, but if I set my network to offline and make a call I get a `Application error: a client-side exception has occurred` with `ERR_INTERNET_DISCONNECTED` in the console. I've also tried try.catch and error.js and no joy. I also tried a custom error boundary and that doesn't catch it - I'm not sure how I can intercept the error.

**Reply by dan-overton** [reply]
I've had the same issue. Bug for it seems to be here but no movement as yet:  https://github.com/vercel/next.js/issues/58638

**Reply by abencun-symphony** [reply]
Same issue here, latest version that works is 14.0.1.

**Comment by tranntuanduong** [comment]
you can change nextjs version to v14.2.0-canary.30 and it work for me.

**Comment by pfurini** [comment]
I'm on 14.2.4 stable and the problem remains. All the examples on next-safe-action.dev are broken too, and they are on Next 15 canary.

I've played around with server actions enough to say that they aren't production ready, period. They can have some use if your production target are public websites with some client interactions, and some features when logged-in, like e-commerce sites or the like. Even for that I will never use server actions for everything but only for simple forms.

For every other app where I need complex forms, client interactions and are data-heavy, like ERPs and dashboards where >90% of the UI is behind authentication, I'll stick with battle-tested react-query and react-hook-form, and route handlers to implement the API endpoints. The added benefit is that customers ask for a native app companion sooner or later, and I can integrate RN alongside the nextjs project in a monorepo, by using Expo or Solito.dev for simpler apps, and my APIs are already there to be consumed by RN.

Maybe this will change in the future, but server actions DX is way worse than the traditional approach, and the drawbacks are more than the benefits.

**Reply by abencun-symphony** [reply]
I have to agree, Server Actions are definitely not fleshed out enough and have also proven to be very unstable implementation-wise since they were introduced as stable. While I like the idea behind them the execution is far from ideal. We went all-in and now we're relying on them less and less as we go forward and keep falling back on React Query with initial server-side data load.

**Reply by bravius** [reply]
The solution I came up with was to dump Next.js and experiment with a "roll your own" framework using Vite, Vike (framework plugin for Vite), and Telefunc.

Telefunc transforms code into functions that work just like React Server Actions. Invoke as client-side async functions, and they execute on the server-side. 

Network errors are catchable, so it solved the essential problem.

Requires trivial server-side middleware to implement.

I know it's a bit extreme to abandon a framework after having invested a lot of time and in it, but sometimes this is necessry to get the results you want. 

https://telefunc.com/

**Comment by ojj1123** [comment]
https://github.com/vercel/next.js/issues/58638
https://github.com/vercel/next.js/pull/61588

This issue is resolved already, isn't it?

I've tried to reproduce this issue, but try/catch can catch the network.

I use this repository and upgrade next.js version to `14.2.5`. It's just working.
(This repository from here https://github.com/vercel/next.js/issues/58638)
https://github.com/4lejandrito/next-server-action-error-issue/

If you have any issue, could you repro?

**Reply by ojj1123** [reply]
@bravius 

Next.js team is preparing the docs on error handling with server actions.
https://github.com/vercel/next.js/discussions/49426#discussioncomment-9970822
